nutils/evaluable.py

```diff
@@ -1629,159 +1631,144 @@ class Multiply(Array):
-        if len(where) != self.ndim:
-            return align(unaligned1 * unaligned2, where, self.shape)
-        for axis1, axis2, *other in map(sorted, func1._diagonals or func2._diagonals):
-            return diagonalize(multiply(takediag(func1, axis1, axis2), takediag(func2, axis1, axis2)), axis1, axis2)
-        for i, parts in func1._inflations:
-            return util.sum(_inflate(f * _take(func2, dofmap, i), dofmap, self.shape[i], i) for dofmap, f in parts.items())
-        for i, parts in func2._inflations:
-            return util.sum(_inflate(_take(func1, dofmap, i) * f, dofmap, self.shape[i], i) for dofmap, f in parts.items())
-        return func1._multiply(func2) or func2._multiply(func1)
+        for j, fj in enumerate(self.funcs):
+            if isinstance(fj, Multiply):
+                return multiply(*self.funcs[:j], *fj.funcs, *self.funcs[j+1:])
+            if fj._const_uniform == 1:
+                return multiply(*self.funcs[:j], *self.funcs[j+1:])
+            for i, parts in fj._inflations:
+                return util.sum(_inflate(multiply(f, *(_take(fi, dofmap, i) for fi in self.funcs[:j] + self.funcs[j+1:])), dofmap, self.shape[i], i) for dofmap, f in parts.items())
+            for axis1, axis2, *other in map(sorted, fj._diagonals):
+                return diagonalize(multiply(*(takediag(f, axis1, axis2) for f in self.funcs)), axis1, axis2)
+            for i, fi in enumerate(self.funcs[:j]):
+                unaligned1, unaligned2, where = unalign(fi, fj)
+                fij = align(unaligned1 * unaligned2, where, self.shape) if len(where) != self.ndim \
+                    else fi._multiply(fj) or fj._multiply(fi)
+                if fij:
+                    return multiply(*self.funcs[:i], *self.funcs[i+1:j], *self.funcs[j+1:], fij)
 
     def _optimized_for_numpy(self):
-        func1, func2 = self.funcs
-        if func1._const_uniform == -1 and func2.dtype != bool:
-            return Negative(func2)
-        if func2._const_uniform == -1 and func1.dtype != bool:
-            return Negative(func1)
-        if self.dtype != complex and func1 == sign(func2):
-            return Absolute(func2)
-        if self.dtype != complex and func2 == sign(func1):
-            return Absolute(func1)
-        if not self.ndim:
-            return
-        unaligned1, where1 = unalign(func1)
-        unaligned2, where2 = unalign(func2)
-        return Einsum((unaligned1, unaligned2), (where1, where2), tuple(range(self.ndim)))
+        for j, fj in enumerate(self.funcs):
+            if fj._const_uniform == -1:
+                return Negative(multiply(*self.funcs[:j], *self.funcs[j+1:]))
+            if isinstance(fj, Sign) and fj.func in self.funcs:
+                fi = fj.func
+                i = self.funcs.index(fi)
+                a, b = sorted([i,j])
+                return multiply(*self.funcs[:a], *self.funcs[a+1:b], *self.funcs[b+1:], Absolute(fi))
```

Comment:
Nice! Why here and not in `_simplified`? Is `fj` never encapsulated in a `Transpose`?

Reply:
Operations like `negative` and `abs` are implemented using basic building blocks to facilitate simplifications, and only during optimization replaced by their `Pointwise` counterparts. This was already the case and this PR is not aiming to change this. The `Transpose` issue is relevant (though rare) but not easily solvable, and the previous implementation suffered the same problem.

Reply:
Perhaps solvable by a property `_unsigned` such that `Sign(func._unsigned)` is equivalent to `func` if not `None` (similar to `_unalign`, `_inflations`, `_diagonals`), but let's leave that for the future.